### PR TITLE
planner: fix a corner case in column pruning rule

### LINKED ISSUE #9125: tidb sql count bug
-- user default test db
use test;

select @@version;
-- 5.7.10-TiDB-v2.1.0-5-g4dad722

CREATE TABLE `t1` (
  `name` varchar(100) DEFAULT NULL
);

insert into t1 (`name`) values ('test');

-- return 0
SELECT COUNT(1)
FROM (
SELECT COUNT(1)
FROM (
SELECT *
FROM t1
WHERE name = 'test') t)
t2;

insert into t1 (`name`) values ('test2');

-- return 1
SELECT COUNT(1)
FROM (
SELECT COUNT(1)
FROM (
SELECT *
FROM t1
WHERE name = 'test') t)
t2;

truncate table t1;

-- repeat above, the seconds count sql return 0;

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
This PR tries to close #9125

### What is changed and how it works?
For a `LogicalAggregation`, if all of its functions are pruned and it doesn't have group-by items, a dummy aggregate function is added to keep the correctness.

### Check List 

Tests 
 - Unit test
 - Integration test

Side effects
 - Possible performance regression

Related changes
 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/rule_column_pruning.go [hunk 1]
```diff
@@ -117,6 +119,17 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
 	for _, aggrFunc := range la.AggFuncs {
 		selfUsedCols = expression.ExtractColumnsFromExpressions(selfUsedCols, aggrFunc.Args, nil)
 	}
+	if len(la.GroupByItems) == 0 && len(la.AggFuncs) == 0 {
+		// If all the aggregate functions are pruned and there is no group-by item, we should add
+		// an aggregate function to keep the correctness.
+		one, err := aggregation.NewAggFuncDesc(la.ctx, ast.AggFuncCount, []expression.Expression{expression.One}, false)
```
Comment (eurekaka) [inline planner/core/rule_column_pruning.go]: I prefer using `first_row(1)` here.

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -117,6 +119,17 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
 	for _, aggrFunc := range la.AggFuncs {
 		selfUsedCols = expression.ExtractColumnsFromExpressions(selfUsedCols, aggrFunc.Args, nil)
 	}
+	if len(la.GroupByItems) == 0 && len(la.AggFuncs) == 0 {
```
Comment (eurekaka) [inline planner/core/rule_column_pruning.go]: If `len(la.GroupByItems) > 0 && len(la.AggFuncs) == 0`, we'd better add this dummy aggregation function as well. Though the result is correct now, we are depending on the specific implementation of executor, if executor is changed in the future, we may have problem for this case then.

File: planner/core/rule_column_pruning.go [hunk 3]
```diff
@@ -117,6 +119,17 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
...
+		// If all the aggregate functions are pruned and there is no group-by item, we should add
+		// an aggregate function to keep the correctness.
+		one, err := aggregation.NewAggFuncDesc(la.ctx, ast.AggFuncCount, []expression.Expression{expression.One}, false)
+		if err != nil {
+			return err
+		}
+		la.AggFuncs = []*aggregation.AggFuncDesc{one}
+		la.schema.Columns = []*expression.Column{{TblName: model.NewCIStr("dummy_cnt"), RetType: types.NewFieldType(mysql.TypeLonglong)}}
```
[the remark this answers is not in the document]
Comment (bb7133) [reply planner/core/rule_column_pruning.go]: Done, thanks

File: planner/core/rule_column_pruning.go (hunk 1 above)
Comment (bb7133) [reply planner/core/rule_column_pruning.go]: You're right... `first_row` is better

File: planner/core/rule_column_pruning.go (hunk 2 above)
Comment (bb7133) [reply planner/core/rule_column_pruning.go]: Thanks for the explaination

**Comment by bb7133** [comment]
addressed, thank you @eurekaka

**Comment by zz-jason** [comment]
needs to be cherry-picked to release-2.1 and release-3.0

**Comment by bb7133** [comment]
> needs to be cherry-picked to release-2.1 and release-3.0

OKay

**Pull request merged**